File: IntegrationTests/Utilities/IntrospectionUtility_ClassTest.cs
// Copyright 2012 rubicon informationstechnologie gmbh
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using InjectionCop.IntegrationTests.Parser;
using InjectionCop.Utilities;
using Microsoft.FxCop.Sdk;
using NUnit.Framework;

namespace InjectionCop.IntegrationTests.Utilities
{
  [TestFixture]
  public class IntrospectionUtility_ClassTest
  {
    [Test]
    public void IsVariable_FieldInMembebinding_IsTrue ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample>("UsingField");
      Block expressionBlock = (Block)sample.Body.Statements[0];
      ExpressionStatement expressionStatement = (ExpressionStatement)expressionBlock.Statements[1];
      UnaryExpression sampleUnaryExpression = (UnaryExpression)expressionStatement.Expression;
      MethodCall sampleMethodCall = (MethodCall) sampleUnaryExpression.Operand;
      Expression sampleExpression = sampleMethodCall.Operands[0];
      bool fieldIsVariable = IntrospectionUtility.IsVariable(sampleExpression);
      Assert.That(fieldIsVariable, Is.True);
    }

    [Test]
    public void IsVariable_Method_IsFalse ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample>("UsingField");
      Block expressionBlock = (Block)sample.Body.Statements[0];
      ExpressionStatement expressionStatement = (ExpressionStatement)expressionBlock.Statements[1];
      UnaryExpression sampleUnaryExpression = (UnaryExpression)expressionStatement.Expression;
      MethodCall sampleMethodCall = (MethodCall) sampleUnaryExpression.Operand;
      bool methodIsVariable = IntrospectionUtility.IsVariable(sampleMethodCall);
      Assert.That(methodIsVariable, Is.False);
    }

    [Test]
    public void GetVariableName_FieldInMembebinding_ReturnsName ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample>("UsingField");
      Block expressionBlock = (Block)sample.Body.Statements[0];
      ExpressionStatement expressionStatement = (ExpressionStatement)expressionBlock.Statements[1];
      UnaryExpression sampleUnaryExpression = (UnaryExpression)expressionStatement.Expression;
      MethodCall sampleMethodCall = (MethodCall) sampleUnaryExpression.Operand;
      Expression sampleExpression = sampleMethodCall.Operands[0];
      string variableName = IntrospectionUtility.GetVariableName(sampleExpression);
      Assert.That(variableName, Is.EqualTo("_field"));
    }

    [Test]
    public void GetVariableName_Method_IsNull ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample>("UsingField");
      Block expressionBlock = (Block)sample.Body.Statements[0];
      ExpressionStatement expressionStatement = (ExpressionStatement)expressionBlock.Statements[1];
      UnaryExpression sampleUnaryExpression = (UnaryExpression)expressionStatement.Expression;
      MethodCall sampleMethodCall = (MethodCall) sampleUnaryExpression.Operand;
      string variableName = IntrospectionUtility.GetVariableName(sampleMethodCall);
      Assert.That(variableName, Is.Null);
    }

    [Test]
    public void IsVariableWithOutParameter_FieldInMemberbinding_IsTrueAndReturnsFieldName ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample>("UsingField");
      Block expressionBlock = (Block)sample.Body.Statements[0];
      ExpressionStatement expressionStatement = (ExpressionStatement)expressionBlock.Statements[1];
      UnaryExpression sampleUnaryExpression = (UnaryExpression)expressionStatement.Expression;
      MethodCall sampleMethodCall = (MethodCall) sampleUnaryExpression.Operand;
      Expression sampleExpression = sampleMethodCall.Operands[0];
      string variableName;
      bool fieldIsVariable = IntrospectionUtility.IsVariable(sampleExpression, out variableName);
      bool correctName = variableName == "_field";
      Assert.That(fieldIsVariable && correctName, Is.True);
    }

    [Test]
    public void IsVariableWithOutParameter_Method_IsFalseAndReturnsNullAsVariableName ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample>("UsingField");
      Block expressionBlock = (Block)sample.Body.Statements[0];
      ExpressionStatement expressionStatement = (ExpressionStatement)expressionBlock.Statements[1];
      UnaryExpression sampleUnaryExpression = (UnaryExpression)expressionStatement.Expression;
      MethodCall sampleMethodCall = (MethodCall) sampleUnaryExpression.Operand;
      string variableName;
      bool methodIsVariable = IntrospectionUtility.IsVariable(sampleMethodCall, out variableName);
      bool correctName = variableName == null;
      Assert.That(!methodIsVariable && correctName, Is.True);
    }

    [Test]
    public void IsField_FieldExpression_ReturnsTrue ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("FieldAssignment");
      Block assignmentBlock = (Block) sample.Body.Statements[0];
      AssignmentStatement assignmentStatement = (AssignmentStatement) assignmentBlock.Statements[1];
      Expression sampleExpression = assignmentStatement.Target;
      bool fieldRecognized = IntrospectionUtility.IsField (sampleExpression);
      Assert.That (fieldRecognized, Is.True);
    }

    [Test]
    public void IsField_NonFieldAssignment_ReturnsFalse ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("NonFieldAssignment");
      Block assignmentBlock = (Block) sample.Body.Statements[0];
      AssignmentStatement assignmentStatement = (AssignmentStatement) assignmentBlock.Statements[1];
      Expression sampleExpression = assignmentStatement.Target;
      bool fieldRecognized = IntrospectionUtility.IsField (sampleExpression);
      Assert.That (fieldRecognized, Is.False);
    }

    [Test]
    public void GetField_FieldExpression_ReturnsContainedField ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("FieldAssignment");
      Block assignmentBlock = (Block) sample.Body.Statements[0];
      AssignmentStatement assignmentStatement = (AssignmentStatement) assignmentBlock.Statements[1];
      Expression sampleExpression = assignmentStatement.Target;
      Field extractedField = IntrospectionUtility.GetField(sampleExpression);

      Member boundMember = ((MemberBinding) sampleExpression).BoundMember;
      bool extractionCorrect = extractedField.UniqueKey == boundMember.UniqueKey;

      Assert.That (extractionCorrect, Is.True);
    }

    [Test]
    public void GetField_NonFieldAssignment_ReturnsNull ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("NonFieldAssignment");
      Block assignmentBlock = (Block) sample.Body.Statements[0];
      AssignmentStatement assignmentStatement = (AssignmentStatement) assignmentBlock.Statements[1];
      Expression sampleExpression = assignmentStatement.Target;
      Field extractedField = IntrospectionUtility.GetField(sampleExpression);
      bool extractionCorrect = extractedField == null;

      Assert.That (extractionCorrect, Is.True);
    }

    [Test]
    public void IsPropertyGetter_PropertyGetter_ReturnsTrue ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("get_AnyProperty");
      bool isPropertyGetter = IntrospectionUtility.IsPropertyGetter(sample);

      Assert.That (isPropertyGetter, Is.True);
    }

    [Test]
    public void IsPropertyGetter_PropertySetter_ReturnsFalse ()
    {
      TypeNode objectTypeNode = IntrospectionUtility.TypeNodeFactory<Object>();
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("set_AnyProperty", objectTypeNode);
      bool isPropertyGetter = IntrospectionUtility.IsPropertyGetter(sample);

      Assert.That (isPropertyGetter, Is.False);
    }

    [Test]
    public void IsPropertyGetter_MethodNamedLikeGetter_ReturnsFalse ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("get_NonExistingProperty");
      bool isPropertyGetter = IntrospectionUtility.IsPropertyGetter (sample);

      Assert.That (isPropertyGetter, Is.False);
    }

    [Test]
    public void IsPropertyGetter_MethodWithParameterNamedLikeGetter_ReturnsFalse ()
    {
      TypeNode stringTypeNode = IntrospectionUtility.TypeNodeFactory<string>();
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("get_NonExistingProperty", stringTypeNode);
      bool isPropertyGetter = IntrospectionUtility.IsPropertyGetter (sample);

      Assert.That (isPropertyGetter, Is.False);
    }

    [Test]
    public void IsPropertyGetter_DummyMethod_ReturnsFalse ()
    {
      TypeNode stringTypeNode = IntrospectionUtility.TypeNodeFactory<string>();
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("Dummy", stringTypeNode);
      bool isPropertyGetter = IntrospectionUtility.IsPropertyGetter (sample);

      Assert.That (isPropertyGetter, Is.False);
    }

    [Test]
    public void IsPropertySetter_PropertySetter_ReturnsTrue ()
    {
      TypeNode objectTypeNode = IntrospectionUtility.TypeNodeFactory<Object>();
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("set_AnyProperty", objectTypeNode);
      bool isPropertyGetter = IntrospectionUtility.IsPropertySetter(sample);

      Assert.That (isPropertyGetter, Is.True);
    }

    [Test]
    public void IsPropertySetter_DummyMethod_ReturnsFalse ()
    {
      TypeNode stringTypeNode = IntrospectionUtility.TypeNodeFactory<string>();
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("Dummy", stringTypeNode);
      bool isPropertySetter = IntrospectionUtility.IsPropertySetter (sample);

      Assert.That (isPropertySetter, Is.False);
    }

    [Test]
    public void IsPropertySetter_PropertyGetter_ReturnsFalse ()
    {
      Method sample = TestHelper.GetSample<IntrospectionUtility_ClassSample> ("get_AnyProperty");
      bool isPropertyGetter = IntrospectionUtility.IsPropertySetter (sample);

      Assert.That (isPropertyGetter, Is.False);
    }
  }
}

File: IntegrationTests.VB/Utilities/IntrospectionUtility_ClassTest.vb
Imports InjectionCop.Utilities
Imports Microsoft.FxCop.Sdk
Imports NUnit.Framework
Imports System
Imports InjectionCop.IntegrationTests.VB.Parser

Namespace Utilities
  <TestFixture()>
  Public Class IntrospectionUtility_ClassTest
    <Test()>
    Public Sub IsVariable_FieldInMembebinding_IsTrue()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingField", New TypeNode() {})
      Dim expressionBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim expressionStatement As ExpressionStatement = CType(expressionBlock.Statements(1), ExpressionStatement)
      Dim sampleUnaryExpression As UnaryExpression = CType(expressionStatement.Expression, UnaryExpression)
      Dim sampleMethodCall As MethodCall = CType(sampleUnaryExpression.Operand, MethodCall)
      Dim sampleExpression As Expression = sampleMethodCall.Operands(0)
      Dim fieldIsVariable As Boolean = IntrospectionUtility.IsVariable(sampleExpression)
      Assert.That(fieldIsVariable, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsVariable_Method_IsFalse()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingField", New TypeNode() {})
      Dim expressionBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim expressionStatement As ExpressionStatement = CType(expressionBlock.Statements(1), ExpressionStatement)
      Dim sampleUnaryExpression As UnaryExpression = CType(expressionStatement.Expression, UnaryExpression)
      Dim sampleMethodCall As MethodCall = CType(sampleUnaryExpression.Operand, MethodCall)
      Dim methodIsVariable As Boolean = IntrospectionUtility.IsVariable(sampleMethodCall)
      Assert.That(methodIsVariable, [Is].[False])
    End Sub

    <Test()>
    Public Sub GetVariableName_FieldInMemberbinding_ReturnsName()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingField", New TypeNode() {})
      Dim expressionBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim expressionStatement As ExpressionStatement = CType(expressionBlock.Statements(1), ExpressionStatement)
      Dim sampleUnaryExpression As UnaryExpression = CType(expressionStatement.Expression, UnaryExpression)
      Dim sampleMethodCall As MethodCall = CType(sampleUnaryExpression.Operand, MethodCall)
      Dim sampleExpression As Expression = sampleMethodCall.Operands(0)
      Dim variableName As String = IntrospectionUtility.GetVariableName(sampleExpression)
      Assert.That(variableName, [Is].EqualTo("_field"))
    End Sub

    <Test()>
    Public Sub GetVariableName_Method_IsNull()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingField", New TypeNode() {})
      Dim expressionBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim expressionStatement As ExpressionStatement = CType(expressionBlock.Statements(1), ExpressionStatement)
      Dim sampleUnaryExpression As UnaryExpression = CType(expressionStatement.Expression, UnaryExpression)
      Dim sampleMethodCall As MethodCall = CType(sampleUnaryExpression.Operand, MethodCall)
      Dim variableName As String = IntrospectionUtility.GetVariableName(sampleMethodCall)
      Assert.That(variableName, [Is].Null)
    End Sub

    <Test()>
    Public Sub IsVariableWithOutParameter_FieldInMemberbinding_IsTrueAndReturnsFieldName()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingField", New TypeNode() {})
      Dim expressionBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim expressionStatement As ExpressionStatement = CType(expressionBlock.Statements(1), ExpressionStatement)
      Dim sampleUnaryExpression As UnaryExpression = CType(expressionStatement.Expression, UnaryExpression)
      Dim sampleMethodCall As MethodCall = CType(sampleUnaryExpression.Operand, MethodCall)
      Dim sampleExpression As Expression = sampleMethodCall.Operands(0)
      Dim variableName As String
      Dim fieldIsVariable As Boolean = IntrospectionUtility.IsVariable(sampleExpression, variableName)
      Dim correctName As Boolean = variableName = "_field"
      Assert.That(fieldIsVariable AndAlso correctName, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsVariableWithOutParameter_Method_IsFalseAndReturnsNullAsVariableName()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingField", New TypeNode() {})
      Dim expressionBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim expressionStatement As ExpressionStatement = CType(expressionBlock.Statements(1), ExpressionStatement)
      Dim sampleUnaryExpression As UnaryExpression = CType(expressionStatement.Expression, UnaryExpression)
      Dim sampleMethodCall As MethodCall = CType(sampleUnaryExpression.Operand, MethodCall)
      Dim variableName As String
      Dim methodIsVariable As Boolean = IntrospectionUtility.IsVariable(sampleMethodCall, variableName)
      Dim correctName As Boolean = variableName = Nothing
      Assert.That(Not methodIsVariable AndAlso correctName, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsField_FieldExpression_ReturnsTrue()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("FieldAssignment", New TypeNode() {})
      Dim assignmentBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim assignmentStatement As AssignmentStatement = CType(assignmentBlock.Statements(1), AssignmentStatement)
      Dim sampleExpression As Expression = assignmentStatement.Target
      Dim fieldRecognized As Boolean = IntrospectionUtility.IsField(sampleExpression)
      Assert.That(fieldRecognized, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsField_NonFieldAssignment_ReturnsFalse()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("NonFieldAssignment", New TypeNode() {})
      Dim assignmentBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim assignmentStatement As AssignmentStatement = CType(assignmentBlock.Statements(1), AssignmentStatement)
      Dim sampleExpression As Expression = assignmentStatement.Target
      Dim fieldRecognized As Boolean = IntrospectionUtility.IsField(sampleExpression)
      Assert.That(fieldRecognized, [Is].[False])
    End Sub

    <Test()>
    Public Sub GetField_FieldExpression_ReturnsContainedField()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("FieldAssignment", New TypeNode() {})
      Dim assignmentBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim assignmentStatement As AssignmentStatement = CType(assignmentBlock.Statements(1), AssignmentStatement)
      Dim sampleExpression As Expression = assignmentStatement.Target
      Dim extractedField As Field = IntrospectionUtility.GetField(sampleExpression)
      Dim boundMember As Member = (CType(sampleExpression, MemberBinding)).BoundMember
      Dim extractionCorrect As Boolean = extractedField.UniqueKey = boundMember.UniqueKey
      Assert.That(extractionCorrect, [Is].[True])
    End Sub

    <Test()>
    Public Sub GetField_NonFieldAssignment_ReturnsNull()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("NonFieldAssignment", New TypeNode() {})
      Dim assignmentBlock As Block = CType(sample.Body.Statements(0), Block)
      Dim assignmentStatement As AssignmentStatement = CType(assignmentBlock.Statements(1), AssignmentStatement)
      Dim sampleExpression As Expression = assignmentStatement.Target
      Dim extractedField As Field = IntrospectionUtility.GetField(sampleExpression)
      Dim extractionCorrect As Boolean = extractedField Is Nothing
      Assert.That(extractionCorrect, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsPropertyGetter_PropertyGetter_ReturnsTrue()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("get_AnyProperty", New TypeNode() {})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertyGetter(sample)
      Assert.That(isPropertyGetter, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsPropertyGetter_PropertySetter_ReturnsFalse()
      Dim objectTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of Object)()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("set_AnyProperty", New TypeNode() {objectTypeNode})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertyGetter(sample)
      Assert.That(isPropertyGetter, [Is].[False])
    End Sub

    <Test()>
    Public Sub IsPropertyGetter_MethodNamedLikeGetter_ReturnsFalse()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("get_NonExistingProperty", New TypeNode() {})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertyGetter(sample)
      Assert.That(isPropertyGetter, [Is].[False])
    End Sub

    <Test()>
    Public Sub IsPropertyGetter_MethodWithParameterNamedLikeGetter_ReturnsFalse()
      Dim stringTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of String)()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("get_NonExistingProperty", New TypeNode() {stringTypeNode})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertyGetter(sample)
      Assert.That(isPropertyGetter, [Is].[False])
    End Sub

    <Test()>
    Public Sub IsPropertyGetter_DummyMethod_ReturnsFalse()
      Dim stringTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of String)()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("Dummy", New TypeNode() {stringTypeNode})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertyGetter(sample)
      Assert.That(isPropertyGetter, [Is].[False])
    End Sub

    <Test()>
    Public Sub IsPropertySetter_PropertySetter_ReturnsTrue()
      Dim objectTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of Object)()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("set_AnyProperty", New TypeNode() {objectTypeNode})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertySetter(sample)
      Assert.That(isPropertyGetter, [Is].[True])
    End Sub

    <Test()>
    Public Sub IsPropertySetter_DummyMethod_ReturnsFalse()
      Dim stringTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of String)()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("Dummy", New TypeNode() {stringTypeNode})
      Dim isPropertySetter As Boolean = IntrospectionUtility.IsPropertySetter(sample)
      Assert.That(isPropertySetter, [Is].[False])
    End Sub

    <Test()>
    Public Sub IsPropertySetter_PropertyGetter_ReturnsFalse()
      Dim sample As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("get_AnyProperty", New TypeNode() {})
      Dim isPropertyGetter As Boolean = IntrospectionUtility.IsPropertySetter(sample)
      Assert.That(isPropertyGetter, [Is].[False])
    End Sub

    <Test()>
    Public Sub GetNestedType_ExistingNestedType_ReturnsNestedType()
      Dim parent As TypeNode = IntrospectionUtility.TypeNodeFactory(Of IntrospectionUtility_ClassSample)()
      Dim nestedType As TypeNode = IntrospectionUtility.GetNestedType(parent, "NestedClass")
      Assert.That(nestedType, [Is].[Not].Null)
    End Sub

    <Test()>
    Public Sub GetNestedType_NonExistingNestedType_ReturnsNestedType()
      Dim parent As TypeNode = IntrospectionUtility.TypeNodeFactory(Of IntrospectionUtility_ClassSample)()
      Dim nestedType As TypeNode = IntrospectionUtility.GetNestedType(parent, "DoesNotExist")
      Assert.That(nestedType, [Is].Null)
    End Sub

    <Test()>
    Public Sub IsVariable_ArrayVariable_ReturnsTrue()
      Dim sampleMethod As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("ArrayVariableAndIndexer", New TypeNode() {})
      Dim sampleBlock As Block = CType(sampleMethod.Body.Statements(0), Block)
      Dim sampleAssignment As AssignmentStatement = CType(sampleBlock.Statements(1), AssignmentStatement)
      Dim sample As Expression = sampleAssignment.Target
      Assert.That(IntrospectionUtility.IsVariable(sample), [Is].[True])
    End Sub

    <Test()>
    Public Sub GetVariableName_ArrayVariable_ReturnsCorrectName()
      Dim sampleMethod As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("ArrayVariableAndIndexer", New TypeNode() {})
      Dim sampleBlock As Block = CType(sampleMethod.Body.Statements(0), Block)
      Dim sampleAssignment As AssignmentStatement = CType(sampleBlock.Statements(1), AssignmentStatement)
      Dim sample As Expression = sampleAssignment.Target
      Assert.That(IntrospectionUtility.GetVariableName(sample), [Is].EqualTo("local$2"))
    End Sub

    <Test()>
    Public Sub IsVariable_Indexer_ReturnsTrue()
      Dim sampleMethod As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("ArrayVariableAndIndexer", New TypeNode() {})
      Dim sampleBlock As Block = CType(sampleMethod.Body.Statements(0), Block)
      Dim sampleAssignment As AssignmentStatement = CType(sampleBlock.Statements(2), AssignmentStatement)
      Dim sample As Expression = sampleAssignment.Target
      Assert.That(IntrospectionUtility.IsVariable(sample), [Is].[True])
    End Sub

    <Test()>
    Public Sub GetVariableName_Indexer_ReturnsNameOfDeclaringObject()
      Dim sampleMethod As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("ArrayVariableAndIndexer", New TypeNode() {})
      Dim sampleBlock As Block = CType(sampleMethod.Body.Statements(0), Block)
      Dim sampleAssignment As AssignmentStatement = CType(sampleBlock.Statements(2), AssignmentStatement)
      Dim sample As Expression = sampleAssignment.Target
      Assert.That(IntrospectionUtility.GetVariableName(sample), [Is].EqualTo("local$2"))
    End Sub

    <Test()>
    Public Sub IsCompilerGenerated_UserGeneratedTypeNode_ReturnsFalse()
      Dim sampleTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of IntrospectionUtility_ClassSample)()
      Assert.That(IntrospectionUtility.IsCompilerGenerated(sampleTypeNode), [Is].[False])
    End Sub

    <Test()>
    Public Sub IsCompilerGenerated_CompilerGeneratedTypeNode_ReturnsTrue()
      Dim sampleMethod As Method = TestHelper.GetSample(Of IntrospectionUtility_ClassSample)("UsingClosure", New TypeNode() {})
      Dim sampleBlock As Block = CType(sampleMethod.Body.Statements(0), Block)
      Dim sampleAssignment As AssignmentStatement = CType(sampleBlock.Statements(1), AssignmentStatement)
      Dim compilerGeneratedTypeNode As TypeNode = sampleAssignment.Source.Type
      Assert.That(IntrospectionUtility.IsCompilerGenerated(compilerGeneratedTypeNode), [Is].[True])
    End Sub
  End Class
End Namespace
